Use SlicerRadiomicsCLI when computing features in SlicerRadiomics module

JoostJM opened the pull request:
Addresses #38 

WIP:
Currently only works when using a parameter file and passing a labelmap. This was a first test to integrate the CLI module.

Add an extra argument in the CLI call to force logging output to level INFO (default WARNING).

cc @Radiomics/developers

JoostJM commented:
@jcfr, I updated the cmake file for building the CLI extension, as I was having trouble getting it to work on Linux (appeared to be a permissions bug). However, I'm not entirely sure if I'm adhering to the CMAKE style and I also don't know if this is necessary for Mac OS and if so, if the current fix also fixes the problem on Mac.

pieper commented on SlicerRadiomicsCLI/SlicerRadiomicsCLI:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python-real
+#!/usr/bin/env python
```
I don't think you want this - python-real will be the slicer-specific one that is installed and will be compatible with the packages in your python path when running in the slicer environment. 

One issue for development is that python-real is created in the install step so it's not convenient to test this in a developer build.

JoostJM replied:
If that's the case then yes, we should keep `python-real`, but is it then possible to also allow for using the `#!/usr/bin/env python` when working with a built slicer (and not an installed one)?

pieper replied:
It might be easiest to make this a shell script that invokes the right python executable depending on the environment.

We want to make sure we use [the python that is installed with all the right settings](https://github.com/Slicer/Slicer/blob/03ef3b63ab8f2c17843b4234cb5d944dedc1e99b/CMake/SlicerBlockInstallPython.cmake).

I can do some tests on mac/linux build vs install trees if needed.

JoostJM replied:
@pieper, I'm currently working on rebuilding the regular module to use the CLI interface, but this is a good point to follow up. I have now also a Linux VM running (Ubuntu) with both a Slicer build and slicer install running. On windows, building Slicer is still giving me headaches from time to time, but there I toy around a lot with a binary installed tree.

pieper replied:
Okay sounds good - let me know if you want help testing anything.

pieper commented:
Yes, the permission change is needed for mac as well.

On Mon, Jan 29, 2018 at 11:50 AM, Joost van Griethuysen <
notifications@github.com> wrote:

> @jcfr <https://github.com/jcfr>, I updated the cmake file for building
> the CLI extension, as I was having trouble getting it to work on Linux
> (appeared to be a permissions bug). However, I'm not entirely sure if I'm
> adhering to the CMAKE style and I also don't know if this is necessary for
> Mac OS and if so, if the current fix also fixes the problem on Mac.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Radiomics/SlicerRadiomics/pull/39#issuecomment-361309218>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHsfYIBZySjPekiT0W8riXIIPpMbL8Iks5tPfbVgaJpZM4RuilR>
> .
>

pieper commented:
Looks like it'll work now - probably easiest to merge it and test installs on all platforms.

JoostJM commented:
> Looks like it'll work now - probably easiest to merge it and test installs on all platforms.

I just want to run a test locally, as I want to make sure I did not build in a infinite loop (where the test is waiting for the script to complete.

JoostJM commented:
@pieper, my suspicions were correct, the while loop that was supposed to wait for the cli to finish blocked the thread, and caused an infinite loop.

I fixed this by reworking the `SlicerRadiomicsLogic` class a bit, so it now also support synchronous extraction using the CLI (by waiting for the CLI to finish inside the logic class). I updated the tests and ran them manually, they passed.

Furthermore, I did a bit of refactoring in the logic class, to better mark functions and variables as 'private', as the class contains a fair amount of these that shouldn't be altered outside of the logic class itself.

pieper commented:
@JoostJM you can make it fully asynchronous and use the observers to tell you when there is progress or the state has changed.  No need for a while loop.

JoostJM commented:
@pieper I did so for regular use, but don't the tests need to be run synchronously? Otherwise the test suite might erroneously think the test has finished, while it's still extracting features. Moreover, the test run 3 extractions, which is not allowed in the user interface (the next can only be started upon completion of the previous). To prevent errors, the tests need to be run one after the other, which only happens when running synronously.

pieper commented:
Okay, I see you were talking about the test code, not just manual testing.  You can have a test that uses an async CLI as linked below but yes, it uses a while loop.

https://github.com/Slicer/Slicer/blob/master/Applications/SlicerApp/Testing/Python/CLIEventTest.py

JoostJM commented:
> Okay, I see you were talking about the test code, not just manual testing. You can have a test that uses an async CLI as linked below but yes, it uses a while loop.

The while loop was indeed in the test code, but I replaced it with a synchronously running extraction. I'll try to see if I can incorporate asynchronous testing as well (following the example you provided).

fedorov commented:
The mac dashboard is failing, looks like it is related to this feature. Is this a known issue?

```
chmod: /.../Radiomics-build/inner-build/lib/Slicer-4.9/cli-modules/./SlicerRadiomicsCLI:
  No such file or directory
```

http://slicer.cdash.org/viewBuildError.php?buildid=1215032
